### PR TITLE
Support Service Binding Spec in  ODO LINK

### PR DESCRIPTION
**Pull request by wtrocki**
## Motivation

Service Binding spec evolved over the last months offering file based bindings. 
This type of bindings are now first class citizens in Quarkus, Spring or Node.js and should be supported by ODO.

Old Env based bindings are still enabled by default so this change is not introducing any breaking/side effects. 

## What was done

This PR adds the following improvements:

1) Updates Service Binding version to 0.5.0 (paving the way for supporting latest changes and features)
2) Adds 2 new flags to enable new features of SBO and alignment with binding specification

```
Flags:
      --bindAsFiles          If present will mount service binding as files in /bindings folder
      --bindingName			 Name of the service binding object
```

Bind as files is defaulting to false for backwards compatibility, but according to the binding spec this should be default to true - it can be added later once we get official documentation/support in all starter apps.

Complete spec can be found here:
https://github.com/k8s-service-bindings/spec

## Verification

1. Use new flags
```
odo link --bindAsFiles=true --bindingName=db  postgresDB
```
Inspect service binding object that was created

**What type of PR is this?**

/kind feature

**What does does this PR do / why we need it**:

General improvements to ODO link command making it work with latest service binding features

**Which issue(s) this PR fixes**:

https://github.com/openshift/odo/issues/4505
https://github.com/openshift/odo/issues/4533
https://github.com/openshift/odo/issues/4534

### CODE REVIEW

**Comment by wtrocki** [comment]
CC @kadel

File: pkg/odo/cli/component/common_link.go [hunk 1]
```diff
@@ -83,6 +85,10 @@ func (o *commonLinkOptions) complete(name string, cmd *cobra.Command, args []str
 		return err
 	}
 
+	if o.bindAsFiles && o.bindingName == "" {
+		return errors.New(`--bindAsFiles option requires --bindingName to be specified`)
```
Comment (wtrocki) [inline pkg/odo/cli/component/common_link.go:89]: I'm not sure if that should be enforced or defaults will work as well

**Comment by maxandersen** [comment]
any reason why bindfiles shouldn't be true by default ?

@kadel @elsony this one would be great to see in !

**Comment by wtrocki** [comment]
@dharmit looking into this now

**Comment by wtrocki** [comment]
```
[ssh:Fedora 32] ++ echo '<<< Need parameter set to minikube or minishift >>>'
[ssh:Fedora 32] ++ exit 1
```

Do I need to setup something

**Comment by dharmit** [comment]
@wtrocki thanks for the PR! There's one problem with the functionality.

When I use `bindAsFiles=true`, `odo push` fails (we need to push after doing `odo link`.) Looking at the events in the pod, I'm seeing:
```sh
$ odo link --bindAsFiles --bindingName=etcd EtcdCluster/example

$ odo push
odo push
 ✓  Waiting for component to start [7ms]

Validation
 ✓  Validating the devfile [29675ns]

Creating Kubernetes resources for component node-todo
 ✗  Waiting for component to start [5m] [WARNING x7: Failed]
 ✗  Failed to start component with name node-todo. Error: Failed to create the component: error while waiting for deployment rollout: timeout while waiting for node-todo deployment roll out\nFor more information to help determine the cause of the error, re-run with '-v'.
See below for a list of failed events that occured more than 5 times during deployment:

 NAME                                         COUNT  REASON  MESSAGE                        

 node-todo-bfdf95cf-6rggw.166e97fc5450b232    7      Failed  Error: secret "etcd" not found 

```
The problem is `Error: secret "etcd" not found`. odo is trying to add a `envFrom` to the pod spec. And it's using the secret `etcd` because when you do `odo link`, it stores link info in `.odo/env/env.yaml`:
```sh
$ cat .odo/env/env.yaml                             
ComponentSettings:
  Name: node-todo
  Project: default
  AppName: app
  Link:
  - Name: etcd    <-----
    ServiceKind: EtcdCluster
    ServiceName: example
  RunMode: run
```
odo tries to find a secret by the same name `etcd` as we used in `odo link`. But it can't find it since SBO created a secret with a different name (Note: earlier SBO used to create a Secret with same name as we passed in the spec for the ServiceBinding):
```sh
$ kubectl get secrets 
NAME                                            TYPE                                  DATA   AGE
etcd-da28a7aa8fd5664bb3177e64c9e4f692b1bd6e1f   Opaque                                1      6m
```
Should SBO create a Secret when using `--bindAsFiles`? Maybe we need to change the spec that we're creating while doing `odo link`? (I'm not sure since I don't know how this feature works.)

Also, this is a problem only when using `bindAsFiles`. If we only provide `bindingName=etcd` or don't provide any arguments proposed in this PR, `odo push` works just fine. So I think it's the files thing that's not functioning as expected.

Also, other than this functionality issue, we will need a few more things to take this PR ahead:
1. Integration tests
2. Validation of the name provided to `--bindingName`.
3. Update `odo link` example in docs under [`docs/public/operator-hub`](https://github.com/openshift/odo/blob/master/docs/public/operator-hub.adoc)

**Comment by dharmit** [comment]
> ```
> [ssh:Fedora 32] ++ echo '<<< Need parameter set to minikube or minishift >>>'
> [ssh:Fedora 32] ++ exit 1
> ```
> 
> Do I need to setup something

@wtrocki Please ignore all the failures for tests with PSI name in it. It's an issue on CI which we're trying to fix.

**Comment by dharmit** [comment]
> Integration tests

Test for `odo link` live under `tests/integration/operatorhub/cmd_service_test.go`.

**Comment by wtrocki** [comment]
I will look into push functionality.
Do you think I can abstract behaviour little bit so link and push will be more aligned?

> Should SBO create a Secret when using --bindAsFiles?

It creates volume in the pod so users would be able to see all values and files.

My take is that when bindAsFiles is provided ODO push needs different handling to detect different way of binding. I will also add integration tests and docs

**Comment by dharmit** [comment]
>  Do you think I can abstract behaviour little bit so link and push will be more aligned?

If user experience and/or overall flow is improving, changing the behaviour should be OK. WDYT @kadel? 
But, what does "more aligned" mean here? 

> > Should SBO create a Secret when using --bindAsFiles?
> 
> It creates volume in the pod so users would be able to see all values and files. Volume is created from secret so this should not differ.

Ah, OK. So a Secret is anytime mandatory. Thanks for explaining ! 👍🏾 

> My take is that when bindAsFiles is provided ODO push needs different handling to detect different way of binding.

Yes. I won't be surprised if you hit some bug or usability issue that's been around since some time but something that didn't get noticed.

**Comment by wtrocki** [comment]
> But, what does "more aligned" mean here?

Ignore me here. Checked push implementation.
Adding some small changes will make this work

**Comment by wtrocki** [comment]
I have been thinking how to refactor this. 
There have been many places where binding is used with assumptions that names are generated and env vars are checked. This needs to be extracted and values from bindings (exact name in CR) instead of computed names should be used. Started doing this but there is obviously a lot of work across different commands for this.

**Comment by wtrocki** [comment]
## Progresss
- [ ] Odo push support
- [ ] Odo unlink support

:)

**Comment by wtrocki** [comment]
I put some thoughts into this. Maybe a form of simplification, we can remove name of the service binding as an option here (it will be actually less complex for end user) while still allowing users to bind things as files. 

When custom name is removed both odo push and unlink will work without any changes, but actual binding will be created with file based links - which is what we looking.

**Comment by wtrocki** [comment]
I'm little bit confused because there are also other service bindings initializers that use kubernetes package instead of sbo: https://github.com/openshift/odo/blob/master/pkg/kclient/serviceCatalog.go#L193-L204

What is the story with this?

**Comment by wtrocki** [comment]
Created https://github.com/openshift/odo/pull/4549 to simplify merge as this kinda becomes little bit priority for me - I would love to have quick chat about the whole approach to custom binding names so we can eventually get both of those options done

**Comment by kadel** [comment]
@wtrocki This PR addresses the Secret name issue https://github.com/openshift/odo/pull/4554

**Comment by mohammedzee1000** [comment]
Please rebase and update your PR to be against new default branch, `main`

**Comment by kadel** [comment]
We can close this PR
I will open a PR to address this soon.
/close

**Pull request closed without merging**